cluster-autoscaler/core/tsc_emulation.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/util/slice"
+)
+
+const (
+	topologySpreadTimeoutAnnotation = "zalando.org/topology-spread-timeout"
+)
+
+type azDistribution map[string]int
+
+func (d azDistribution) addPod(zoneName string) {
+	d[zoneName] += 1
+}
+
+func (d azDistribution) smallestZone(availableZones []string) string {
+	currentCandidate := ""
+	currentCount := 0
+	for _, zone := range availableZones {
+		zoneCount := d[zone]
+		if currentCandidate == "" || zoneCount < currentCount {
+			currentCandidate = zone
+			currentCount = zoneCount
+		}
+	}
+	return currentCandidate
+}
+
+type azDistributions map[string]azDistribution
+
+func (d azDistributions) distributionFor(distributionId string) azDistribution {
+	result := d[distributionId]
+	if result == nil {
+		result = azDistribution{}
+		d[distributionId] = result
+	}
+	return result
+}
+
+type topologySpreadEmulationScheduledPodLister struct {
+	base            kubernetes.PodLister
+	constraintLabel string
+	nodeZoneMapping map[string]string
+}
+
+func topologySpreadEmulationWrapScheduledPodLister(base kubernetes.PodLister, options config.AutoscalingOptions, nodeZoneMapping map[string]string) kubernetes.PodLister {
+	if options.EmulatedTopologySpreadConstraintLabel == "" {
+		return base
+	}
+	return &topologySpreadEmulationScheduledPodLister{
+		base:            base,
+		constraintLabel: options.EmulatedTopologySpreadConstraintLabel,
+		nodeZoneMapping: nodeZoneMapping,
+	}
+}
+
+func (l *topologySpreadEmulationScheduledPodLister) List() ([]*corev1.Pod, error) {
+	original, err := l.base.List()
+	if err != nil {
+		return nil, err
+	}
+
+	var result []*corev1.Pod
+	for _, pod := range original {
+		if hasEmulatedTopologySpreadConstraints(pod, l.constraintLabel) {
+			pod = pod.DeepCopy()
+			pod.Spec.TopologySpreadConstraints = nil
+
+			// Pin already scheduled pods to their current zones, to avoid potential issues with downscaling
+			if nodeZone, ok := l.nodeZoneMapping[pod.Spec.NodeName]; ok && nodeZone != "" {
```

Comment:
Do we consider `Succeed/Failed` pods here? I believe those also have the nodeName set which could result in wrong distribution if we don't filter them out somewhere else?

Reply:
This wraps a [`scheduledPodLister`](https://github.com/zalando-incubator/autoscaler/blob/0e73b3e53035bfb05d0af1290d099895fe121dc7/cluster-autoscaler/utils/kubernetes/listers.go#L207), which already filters those out.

Reply:
(no idea what's going on with the selector there, don't ask!)